### PR TITLE
feat: Add checker to ensure that env variable doc is up to date

### PR DESCRIPTION
**Pull request by simster7**
Checklist:

* [ ] My organization is added to [USERS.md](https://github.com/argoproj/argo-workflows/blob/master/USERS.md).

### CODE REVIEW

File: hack/check-env-doc.sh [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+
+cat ./docs/environment-variables.md \
+  | grep "| \`" \
+  | awk '{gsub(/\`/, "", $2);  print $2; }' \
+  | while read -r x; do
+    if ! grep -qR $x ./workflow ./persist ./util; then
+      echo "Documented variable $x is not used anywhere";
+      exit 1;
+    fi;
+  done
```
Comment (simster7) [inline hack/check-env-doc.sh:11]: Need to add reverse case: "Variable in use is not documented"

**Review by terrytangyuan — COMMENTED** [review]
This is great to have! Thanks. Would you like to get https://github.com/argoproj/argo-workflows/pull/5080 merged so that we can test out the script in this PR?

File: docs/environment-variables.md [hunk 2]
```diff
@@ -23,6 +23,8 @@ Note that these environment variables may be removed at any time.
 | `TRANSIENT_ERROR_PATTERN` | `string` | The regular expression that represents additional patterns for transient errors. |
 | `WF_DEL_PROPAGATION_POLICY` | `string` | The deletion propogation policy for workflows. |
 | `WORKFLOW_GC_PERIOD` | `time.Duration` | The periodicity for GC of workflows. |
+| `BUBBLE_ENTRY_TEMPLATE_ERR` | `bool` | Whether to bubble up template errors to workflow. Default true |
+| `INFORMER_WRITE_BACK` | `bool` | Whether to write back to informer instead of catching up. Deafult true |
```
Comment (simster7) [inline docs/environment-variables.md:27]: @terrytangyuan Caught a couple you missed :)

Comment (terrytangyuan) [reply docs/environment-variables.md:27]: Nice! 👍

File: hack/check-env-doc.sh (hunk 1 above)
Comment (simster7) [reply hack/check-env-doc.sh]: This is done

File: hack/check-env-doc.sh [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+echo "Checking env variables doc for completness..."
+
+function check-used {
+    grep "| \`" < ./docs/environment-variables.md \
+      | awk '{gsub(/\`/, "", $2);  print $2; }' \
+      | while read -r x; do
+        var="${x%\`}";
+        var="${var#\`}";
+        if ! grep -qR --exclude="*_test.go" "$var" ./workflow ./persist ./util; then
+          echo "Documented variable $var is not used anywhere";
```
Comment (terrytangyuan) [inline hack/check-env-doc.sh:12]:
```suggestion
          echo "Documented variable $var in docs/environment-variables.md is not used anywhere";
```

File: hack/check-env-doc.sh [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+function check-documented {
+    grep -REh --exclude="*_test.go" "Getenv.*?\(|LookupEnv.*?\(" ./workflow ./persist ./util \
+      | grep -Eo "\"[A-Z_]+?\"" \
+      | sort \
+      | uniq \
+      | while read -r x; do
+        var="${x%\"}";
+        var="${var#\"}";
+        if ! grep -q "$var" docs/environment-variables.md; then
+          echo "Variable $var not documented in docs/environment-variables";
```
Comment (terrytangyuan) [inline hack/check-env-doc.sh:27]:
```suggestion
          echo "Variable $var not documented in docs/environment-variables.md";
```

File: hack/check-env-doc.sh [hunk 5]
```diff
@@ -0,0 +1,33 @@
+#!/bin/bash
+
+echo "Checking env variables doc for completness..."
```
Comment (terrytangyuan) [inline hack/check-env-doc.sh:3]: Perhaps be explicit about where the file is

Comment (terrytangyuan) [inline hack/check-env-doc.sh:3]:
```suggestion
echo "Checking docs/environment-variables.md for completeness..."
```

**Review by terrytangyuan — COMMENTED** [review]
Looks great! Just some small nitpicks.

File: hack/check-env-doc.sh [hunk 6]
```diff
@@ -0,0 +1,33 @@
+#!/bin/bash
+
+echo "Checking docs/environment-variables.md for completeness..."
+
+function check-used {
```
Comment (simster7) [inline hack/check-env-doc.sh:5]: Add `-o`

**Review by alexec — APPROVED** [review]

**Pull request merged**